kong/clustering/version_negotiation/init.lua

```diff
@@ -0,0 +1,340 @@
+  return {
+    id = kong.node.get_id()
+  }
+end
+
+local function cp_priority(name, req_versions, known_versions)
+  local versions_set = {}
+  for _, version in ipairs(req_versions) do
+    versions_set[version] = true
+  end
+
+  for _, v in ipairs(known_versions) do
+    if versions_set[v.version] then
+      return true, {
+        name = name,
+        version = v.version,
+        message = v.message,
+      }
+    end
+  end
+
+  return false, { name = name, message = "No valid version" }
+end
+
+local all_known_services = require "kong.clustering.version_negotiation.services_known"
+
+local function check_node_compatibility(client_node)
+  if client_node.type ~= "KONG" then
+    return nil, ("unknown node type %q"):format(client_node.type), CLUSTERING_SYNC_STATUS.UNKNOWN
+  end
+
+  local ok, msg, status = clustering_utils.check_kong_version_compatibility(KONG_VERSION, client_node.version)
+  if not ok then
+    return ok, msg, status
+  end
+
+  return true, nil, CLUSTERING_SYNC_STATUS.NORMAL
+end
+
+local function do_negotiation(req_body)
+  local services_accepted = {}
+  local services_rejected = {}
+
+  for _, req_service in ipairs(req_body.services_requested) do
+    if type(req_service) ~= "table" or type(req_service.name) ~= "string" then
+      goto continue
+    end
+
+    local name = req_service.name
+
+    if type(req_service.versions) ~= "table" then
+      table.insert(services_rejected, {
+        name = name,
+        message = "invalid \"versions\" table.",
+      })
+      goto continue
+    end
+
+    local known_service = all_known_services[name]
+    if not known_service then
+      table.insert(services_rejected, {
+        name = name,
+        message = "unknown service.",
+      })
+      goto continue
+    end
+
+    local ok, service_response = cp_priority(name, req_service.versions, known_service)
+    if ok then
+      ngx_log(ngx_DEBUG, _log_prefix,
+        "accepted: \"" .. service_response.name ..
+        "\", version \"" .. service_response.version ..
+        "\": ".. service_response.message)
+      table.insert(services_accepted, service_response)
+    else
+
+      ngx_log(ngx_DEBUG, _log_prefix,
+        "rejected: \"" .. service_response.name ..
+        "\": " .. service_response.message)
+      table.insert(services_rejected, service_response)
+    end
+
+    ::continue::
+  end
+
+  return {
+    node = node_info(),
+    services_accepted = services_accepted,
+    services_rejected = services_rejected,
+  }
+end
+
+
+local function register_client(conf, client_node, services_accepted)
+  local services_map = {}
+  for _, service in ipairs(services_accepted) do
+    services_map[service.name] = service.version
+  end
+
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = client_node.id, }, {
+    last_seen = ngx.time(),
+    config_hash = "00000000000000000000000000000000",
+    hostname = client_node.hostname,
+    ip = ngx.var.realip_remote_addr or ngx.var.remote_addr,
```

Comment:
Any particular reason for using `$realip_remote_addr` here? I would think we would want to play nicely with a control plane behind a reverse proxy, no?

Reply:
apparently i copied that from the `pdk.client.get_ip()` function.  Yes, control planes behind a reverse proxy are definitely a thing, so I'd think using the value provided by the proxy would be better than what the socket sees.  Do I have it backwards?

Reply:
Gotcha, yeah `$realip_remote_addr` will always contain the actual IP of the immediate downstream/client (as opposed to what is derived from X-Forwarded-For), so I think `$remote_addr` is all you need here:

```lua
ip = ngx.var.remote_addr,
```

Reply:
right, it took me several readings of the nginx docs, but i think i get it now, the "real ip" module sets the "normal" variables using proxy info, but stores the "original" ip in the `real_ip...` vars.   kinda using "real" word as meaning both the "real origin" and the "original origin".   big mess.